### PR TITLE
Simplify `get_crossmatchted_antibodies`

### PR DESCRIPTION
**Pull request by jogobeny**
Hello there, I've
1. divided the crossmatching into seperated types
2. added https://github.com/mild-blue/txmatching/blob/4f4dcb0b00f258405084d7837628d3aaac648b0f/txmatching/utils/hla_system/hla_crossmatch.py#L125 because of not getting antibodies below cutoff in https://github.com/mild-blue/txmatching/blob/4f4dcb0b00f258405084d7837628d3aaac648b0f/txmatching/utils/hla_system/hla_crossmatch.py#L142-L145
3. removed https://github.com/mild-blue/txmatching/blob/06f545e00f9eb31fb0c46a79b73d87cf0c6584cf/txmatching/utils/hla_system/hla_crossmatch.py#L82 from the logic code and modified https://github.com/mild-blue/txmatching/blob/06f545e00f9eb31fb0c46a79b73d87cf0c6584cf/tests/web/test_patient_api.py#L523 to https://github.com/mild-blue/txmatching/blob/4f4dcb0b00f258405084d7837628d3aaac648b0f/tests/web/test_patient_api.py#L523 to check the uniqueness of antibodies.

What are you thoughts?

### CODE REVIEW

**Review by kristinagalik — DISMISSED** [review]
fine by me :) so when @kubantjan is back he can check it too 

(also rebase)

File: tests/web/test_patient_api.py [hunk 1]
```diff
@@ -564,9 +567,13 @@ def test_sorted_antibodies(self):
                              headers=self.auth_headers)
         self.assertEqual(200, res.status_code)
 
+        # Because there is a uniqueness between antibodies in high resolution and
```
Comment (kubantjan) [inline tests/web/test_patient_api.py:570]: https://stackoverflow.com/questions/480214/how-do-i-remove-duplicates-from-a-list-while-preserving-order

podle toho co tam pisou dava smysl list(dict.fromkeys(items))

a tim bych i zjednodusil ten komentar na neco jako

using code from link to get unique elements from list while preserving order.

File: txmatching/utils/hla_system/hla_crossmatch.py [hunk 2]
```diff
@@ -53,96 +53,100 @@ def _get_antibodies_with_high_res(antibodies: List[HLAAntibody]) -> List[HLAAnti
     return [antibody for antibody in antibodies if antibody.code.high_res]
 
 
-# TODO improve and remove pylint ignores https://github.com/mild-blue/txmatching/issues/947
-# pylint: disable=too-many-branches,too-many-locals,too-many-nested-blocks,too-many-statements
+def _do_crossmatch(matching_antibodies: List[HLAAntibody], positive_matches: Set[AntibodyMatch],
```
Comment (kubantjan) [inline txmatching/utils/hla_system/hla_crossmatch.py:56]: co maji znamenat ty podrzitka ve jmenech tech parametru metody?

**Review by kubantjan — CHANGES_REQUESTED** [review]
jeste in progress

File: txmatching/utils/hla_system/hla_crossmatch.py [hunk 3]
```diff
@@ -53,96 +53,100 @@ def _get_antibodies_with_high_res(antibodies: List[HLAAntibody]) -> List[HLAAnti
...
+                   _with_type: AntibodyMatchTypes, _type: AntibodyMatchTypes) -> bool:
+    if len(matching_antibodies) <= 0:
+        return False
+
+    antibodies_over_cutoff = _get_antibodies_over_cutoff(matching_antibodies)
+    antibodies_with_high_res = _get_antibodies_with_high_res(matching_antibodies)
+    if set(matching_antibodies) == set(antibodies_over_cutoff) == set(antibodies_with_high_res):
+        for antibody_over_cutoff in antibodies_over_cutoff:
+            positive_matches.add(AntibodyMatch(antibody_over_cutoff, AntibodyMatchTypes.HIGH_RES))
+    elif len(_get_antibodies_over_cutoff(antibodies_with_high_res)) > 0:
```
Comment (kubantjan) [inline txmatching/utils/hla_system/hla_crossmatch.py:66]: tohle je takovy divny, kdyz bude nejaky antibody high res a nejaky ne, tak tady udelas to ze vsem das to high_res_with_broad, coz bude spatne.

tady spis bys mel projet ten seznam tech matchu a u kazdyho z nich se rozhodnout, podle toho, jestli je to high res nebo ne.

**Review by kubantjan — CHANGES_REQUESTED** [review]
par komentaru

**Comment by kubantjan** [comment]
Mrkni v ramci tohoto na:

https://github.com/mild-blue/txmatching/pull/984

asi by bylo idealni v ramci toho upravit tu logiku, aby to odpovidalo presne tomuhle (+ pridat testy) 

tim se ozkousi, jestli je to napsane srozumitelne 

v  zasade by to cca takto uz ted melo fungovat, ale ted je to hezceji shrnute a asi mi i dava smysl aby ten kod mel takovouhle vyhybku. Tzn ten mode A a mode B

**Comment by kubantjan** [comment]
Lepsi ty komentare nedavat do todo ale rovou do toho PR, pak se na ne lip odpovida

File: tests/utils/hla_system/test_crossmatch.py [hunk 4]
```diff
@@ -224,73 +246,168 @@ def test_antibodies_with_multiple_mfis(self):
...
+        self._assert_matches_equal('A23',
                                    [create_antibody('A*23:01', 2100, 2000),
-                                    create_antibody('A*23:01', 2200, 2000)], True,
-                                   [AntibodyMatch(create_antibody('A*23:01', 2150, 2000), AntibodyMatchTypes.HIGH_RES)])
+                                    create_antibody('A*23:04', 1900, 2000)], True,
+                                   [AntibodyMatch(create_antibody('A*23:01', 2100, 2000),
+                                                  AntibodyMatchTypes.HIGH_RES_WITH_SPLIT)],
+                                   is_type_a=True)
+
+        # TODO: Je toto potřeba? `create_antibodies` stejně volá nějaké parsování,
```
Comment (kubantjan) [inline tests/utils/hla_system/test_crossmatch.py:330]: neni potreba, mas pravdu muzes smazat

File: tests/utils/hla_system/test_crossmatch.py [hunk 5]
```diff
@@ -224,73 +246,168 @@ def test_antibodies_with_multiple_mfis(self):
...
+        # first matching antibody with mfi < cutoff, second with mfi > cutoff  # BROAD_1
         self._assert_matches_equal('A9',
                                    [create_antibody('A9', 1900, 2000),
                                     create_antibody('A9', 2100, 2000)], True,
-                                   [AntibodyMatch(create_antibody('A9', 2000, 2000), AntibodyMatchTypes.BROAD)])
-        # first matching antibody with mfi1 > cutoff, second with mfi2 > mfi1
+                                   [AntibodyMatch(create_antibody('A9', 2000, 2000), AntibodyMatchTypes.BROAD)],
+                                   is_type_a=False)
+
+        # TODO: Tohle už je taky podle mě zbytečné, protože to, že se správně počítá průměr mfi jsme zjistili už nahoře
```
Comment (kubantjan) [inline tests/utils/hla_system/test_crossmatch.py:359]: jop, smaz

File: tests/utils/hla_system/test_crossmatch.py [hunk 6]
```diff
@@ -179,36 +189,48 @@ def test_get_crossmatched_antibodies(self):
...
+                                                  AntibodyMatchTypes.HIGH_RES)],
+                                   is_type_a=False)
+        # positive split crossmatch  # HIGH_RES_2
         self._assert_matches_equal('A*23:01', [create_antibody('A*23:04', 2100, 2000)], True,
                                    [AntibodyMatch(create_antibody('A*23:04', 2100, 2000),
-                                                  AntibodyMatchTypes.HIGH_RES)])
+                                                  AntibodyMatchTypes.HIGH_RES)],
+                                   is_type_a=True)
         # negative split crossmatch
+        # TODO: Tohle mi podle té logicky odpovídá pro type B typizací BROAD, podle BROAD_2.
```
Comment (kubantjan) [inline tests/utils/hla_system/test_crossmatch.py:203]: Pro obe varianty to ma byt NONE. A podle me to i odpovida popisu v textu.

Pro "b" , aby to byl broad, muselo by byt bud donoruv antiben, nebo recipientovo antibody v broad, coz v tomto pripade neni.

Pro "a" aby to byl HIGH_2, muselo by jit o shodu na urovni split a ne broad.

File: tests/utils/hla_system/test_crossmatch.py [hunk 7]
```diff
@@ -224,73 +246,168 @@ def test_antibodies_with_multiple_mfis(self):
...
 
-        # first matching antibody with mfi > cutoff, second with mfi < cutoff
-        self._assert_matches_equal('A23',
-                                   [create_antibody('A*23:01', 2100, 2000),
-                                    create_antibody('A*23:04', 1900, 2000)], True,
-                                   [AntibodyMatch(create_antibody('A*23:01', 2100, 2000),
-                                                  AntibodyMatchTypes.HIGH_RES_WITH_SPLIT)])
-        # first matching antibody with mfi < cutoff, second with mfi > cutoff
-        self._assert_matches_equal('A23',
+        # antibodies duplicity should not raise duplicity assert, because the antibodies are joined before creating
```
Comment (kubantjan) [inline tests/utils/hla_system/test_crossmatch.py:251]: tohleto s tim duplicity assertem nevim proc tady pises?

**Review by kubantjan — CHANGES_REQUESTED** [review]
komentare

obecne bych jeste udelal 2 3 testy na _is_recipient_type_a

File: txmatching/utils/hla_system/hla_crossmatch.py [hunk 8]
```diff
@@ -53,97 +54,196 @@ def _get_antibodies_with_high_res(antibodies: List[HLAAntibody]) -> List[HLAAnti
...
+
+    for antibodies_per_group in recipient_antibodies.hla_antibodies_per_groups:
+        total_antibodies += len(antibodies_per_group.hla_antibody_list)
+        for hla_antibody in antibodies_per_group.hla_antibody_list:
+            if hla_antibody.code.high_res is None:
+                return False
+            if hla_antibody.mfi < hla_antibody.cutoff:
+                is_at_least_one_antibody_below_cutoff = True
+
+    if total_antibodies < 5:
```
Comment (kubantjan) [inline txmatching/utils/hla_system/hla_crossmatch.py:229]: tuhle konstantu dej nahoru do filu pojmenovanou. Nikdy nechces mit v kodu magic numbers. a ma byt 20 a ne 5 :)

File: txmatching/utils/hla_system/hla_crossmatch.py [hunk 9]
```diff
@@ -53,97 +54,196 @@ def _get_antibodies_with_high_res(antibodies: List[HLAAntibody]) -> List[HLAAnti
...
+                        positive_matches.append(AntibodyMatch(match, AntibodyMatchTypes.BROAD))
+                    continue
+
+        _add_none_typization(antibodies, positive_matches)
+        antibody_matches_for_groups.append(AntibodyMatchForHLAGroup(hla_per_group.hla_group, positive_matches))
+
+    return antibody_matches_for_groups
+
+
+def _is_recipient_type_a(recipient_antibodies: HLAAntibodies) -> bool:
```
Comment (kubantjan) [inline txmatching/utils/hla_system/hla_crossmatch.py:217]: logika ok

**Review by kubantjan — CHANGES_REQUESTED** [review]

File: txmatching/utils/hla_system/hla_crossmatch.py [hunk 10]
```diff
@@ -53,97 +54,196 @@ def _get_antibodies_with_high_res(antibodies: List[HLAAntibody]) -> List[HLAAnti
     return [antibody for antibody in antibodies if antibody.code.high_res]
```
Comment (kristinagalik) [inline txmatching/utils/hla_system/hla_crossmatch.py:54]: this function is no longer used

File: txmatching/utils/hla_system/hla_crossmatch.py [hunk 11]
```diff
@@ -53,97 +54,196 @@ def _get_antibodies_with_high_res(antibodies: List[HLAAntibody]) -> List[HLAAnti
...
+        for hla_antibody in antibodies_per_group.hla_antibody_list:
+            if hla_antibody.code.high_res is None:
+                return False
+            if hla_antibody.mfi < hla_antibody.cutoff:
+                is_at_least_one_antibody_below_cutoff = True
+
+    if total_antibodies < 5:
+        return False
+
+    if is_at_least_one_antibody_below_cutoff:
```
Comment (kristinagalik) [inline txmatching/utils/hla_system/hla_crossmatch.py:232]: shouldnt this be `if not is_at_least_one_antibody_below_cutoff`?

**Review by kristinagalik — CHANGES_REQUESTED** [review]
(ill finish review tomorrow in the afternoon)

**Comment by kubantjan** [comment]
> Okay, I've
> 
> 1. fixed mistakes in the logic in `get_crossmatched_antibodies`: [Simplify `get_crossmatchted_antibodies` #971 (comment)](https://github.com/mild-blue/txmatching/pull/971#discussion_r978343515)
> 2. added `test_is_recipient_type_a`
> 3. made to pass all tests for `get_crossmatched_antibodies` function
> 
> I have a question about `test_virtual_crossmatch_if_all_antibodies_positive_in_high_res`. I am not sure about determining whether a crossmatch is positive or negative. For example `self._assert_negative_crossmatch('A9', high_res_antibodies_not_all_positive, True, HLACrossmatchLevel.BROAD)` should have negative crossmatch on `BROAD` level. This assert calls `is_positive_hla_crossmatch` -> `get_crossmatched_antibodies`, but in lack of antibodies, it always does logic type B (and I don't even know if it should be type B or type A). For type B, this crossmatch is positive, because antibodies has `BROAD` match type and `BROAD` is accepted in `BROAD` crossmatch level. For type A, the crossmatch is negative, because antibodies has `HIGH_RES_WITH_BROAD` match type and it is not accepted by `BROAD` crossmatch level. I mean, I've shown you that it is determinable if it's type A or type B by more analyzing, but I'd like to have it approved by you. I might have missed some text in documentation.

cely ten test ma byt pro logiku type A

File: txmatching/utils/hla_system/hla_crossmatch.py [hunk 12]
```diff
@@ -49,101 +53,203 @@ def _get_antibodies_over_cutoff(antibodies: List[HLAAntibody]) -> List[HLAAntibo
...
-                        for antibody_over_cutoff in antibodies_over_cutoff:
-                            positive_matches.add(
-                                AntibodyMatch(antibody_over_cutoff, AntibodyMatchTypes.HIGH_RES_WITH_SPLIT))
-                    else:
-                        for antibody_over_cutoff in antibodies_over_cutoff:
-                            positive_matches.add(AntibodyMatch(antibody_over_cutoff, AntibodyMatchTypes.SPLIT))
+                tested_antibodies_that_match = [antibody for antibody in antibodies
+                                                if hla_type.code.split == antibody.code.split]
+                if len(tested_antibodies_that_match) > 0:
+                    for match in _get_antibodies_over_cutoff(tested_antibodies_that_match):
```
Comment (kubantjan) [inline txmatching/utils/hla_system/hla_crossmatch.py:199]: v celym antibodies type b te zajimaji jen over cutoff, tak bych to udelal rovnou na zacatku.

File: txmatching/utils/hla_system/hla_crossmatch.py [hunk 13]
```diff
@@ -49,101 +53,203 @@ def _get_antibodies_over_cutoff(antibodies: List[HLAAntibody]) -> List[HLAAntibo
     return [antibody for antibody in antibodies if antibody.mfi >= antibody.cutoff]
 
 
-def _get_antibodies_with_high_res(antibodies: List[HLAAntibody]) -> List[HLAAntibody]:
-    return [antibody for antibody in antibodies if antibody.code.high_res]
+def _add_undecidable_typization(antibodies: List[HLAAntibody], hla_per_group: HLAPerGroup,
```
Comment (kubantjan) [inline txmatching/utils/hla_system/hla_crossmatch.py:56]: sem by mely vstupovat jen pozitivni protilatky. coz se ted nedeje

**Review by kubantjan — CHANGES_REQUESTED** [review]
Crossmatch by mel vracet rozhodne vzdy a za kazdych okolnosti jen list protilatek over cutoff. Ty pod cutoff slouzi jen jako doprovodna informace pro rozhodnuti do ktery kategorie ta dana protilatka nad cutoff spada. Coz ted tak uplne mi neprijde (treba u toho add_undecidable)

**Review by kubantjan — COMMENTED** [review]
Jinak celkove vypada dobre

**Comment by kubantjan** [comment]
> 

None urcite nikdy nic matchnout nemuze

**Review by kubantjan — CHANGES_REQUESTED** [review]
Tady porad cekam na tebe ne?

**Review by kristinagalik — COMMENTED** [review]
odomna len maly dotaz

File: txmatching/utils/hla_system/hla_crossmatch.py [hunk 14]
```diff
@@ -49,108 +234,95 @@ def _get_antibodies_over_cutoff(antibodies: List[HLAAntibody]) -> List[HLAAntibo
...
+                                       positive_matches):
+        return True
+
+    return False
+
+
+def _add_undecidable_typization(antibodies: List[HLAAntibody],
+                                hla_per_group: HLAPerGroup,
+                                positive_matches: List[AntibodyMatch]):
+    if hla_per_group.hla_group == HLAGroup.Other:
```
[the remark this answers is not in the document]
Comment (kubantjan) [reply txmatching/utils/hla_system/hla_crossmatch.py:316]: https://www.quora.com/Are-function-calls-expensive-in-Python

myslim ze tohle je naprosto ok, kdyz bys delal nejakou hlubokou rekurzi, tak to dava asmysl resit, tady v tomto pripade ne.

File: txmatching/utils/hla_system/hla_crossmatch.py [hunk 15]
```diff
@@ -18,11 +21,192 @@ class AntibodyMatchForHLAGroup:
...
+        antibodies_per_hla_group.antibody_matches = sorted(
+            antibodies_per_hla_group.antibody_matches,
+            key=lambda hla_group: (
+                hla_group.hla_antibody.raw_code,
+            ))
+
+    return antibody_matches_for_groups
+
+
 def is_positive_hla_crossmatch(donor_hla_typing: HLATyping,
```
Comment (kubantjan) [inline txmatching/utils/hla_system/hla_crossmatch.py:205]: tohle je druha hlavni, hned pod ni.

File: txmatching/utils/hla_system/hla_crossmatch.py [hunk 16]
```diff
@@ -49,108 +234,95 @@ def _get_antibodies_over_cutoff(antibodies: List[HLAAntibody]) -> List[HLAAntibo
...
-# TODO improve and remove pylint ignores https://github.com/mild-blue/txmatching/issues/947
-# pylint: disable=too-many-branches,too-many-locals,too-many-nested-blocks,too-many-statements
-def get_crossmatched_antibodies(donor_hla_typing: HLATyping,
-                                recipient_antibodies: HLAAntibodies,
-                                use_high_resolution: bool) -> List[AntibodyMatchForHLAGroup]:
-    antibody_matches_for_groups = []
-    for hla_per_group, antibodies_per_group in zip(donor_hla_typing.hla_per_groups,
-                                                   recipient_antibodies.hla_antibodies_per_groups):
-        assert hla_per_group.hla_group == antibodies_per_group.hla_group
+def _add_all_tested_antibodies_are_positive(tested_antibodies_that_match: List[HLAAntibody],
```
Comment (kubantjan) [inline txmatching/utils/hla_system/hla_crossmatch.py:241]: tenhle nazev mi prijde blbe srozumitelny.

co to presne dela? Z nazvu tomu uprimne nerozumim a zkodu si taky nejsem jisty.

File: txmatching/utils/hla_system/hla_crossmatch.py [hunk 17]
```diff
@@ -18,11 +21,192 @@ class AntibodyMatchForHLAGroup:
...
+            if use_high_resolution and hla_type.code.high_res is not None:
+                tested_antibodies_that_match = [antibody for antibody in antibodies
+                                                if hla_type.code.high_res == antibody.code.high_res]
+                # HIGH_RES_1
+                if _add_positive_tested_antibodies(tested_antibodies_that_match,
+                                                   AntibodyMatchTypes.HIGH_RES,
+                                                   positive_matches):
+                    continue
+
+            if hla_type.code.split is not None:
```
Comment (kubantjan) [inline txmatching/utils/hla_system/hla_crossmatch.py:141]: odkaz na readme

**Review by kubantjan — CHANGES_REQUESTED** [review]
Pridej vysvetleni vyznam co to znamena (tzn HIGH_RES_1 ze odkazuje na bod 1. z toho atp. Asi to dej tohle vysvetleni do readme..

File: txmatching/utils/hla_system/hla_crossmatch.py (hunk 17 above)
Comment (kubantjan) [reply txmatching/utils/hla_system/hla_crossmatch.py:141]: SPLIT_1 asi ne?

**Comment by kubantjan** [comment]
Uvedomil jsem si chybu.

V pripade SPLIT_2 a BROAD_2 kdyz je antigen v high res, tak ten bod jeplatny jen pro type B. Pro type A se tohle resi uz jinde (bud to bude HIGH_RES, nebo to bude HIGH_RES_WITH_SPLIT/BROAD

mrkni na to prosim, pripadne pridej test a fix.

**Review by kubantjan — APPROVED** [review]

**Pull request merged**